cohere/client.py

```diff
@@ -465,61 +469,40 @@ def __print_warning_msg(self, response: Response):
         if 'X-API-Warning' in response.headers:
             print("\033[93mWarning: {}\n\033[0m".format(response.headers['X-API-Warning']), file=sys.stderr)
 
-    def __pyfetch(self, url, headers, json_body, method='POST') -> Response:
-        req = XMLHttpRequest.new()
-        req.open(method, url, False)
-        for key, value in headers.items():
-            req.setRequestHeader(key, value)
-        try:
-            req.send(json_body)
-        except Exception:
-            raise CohereError(message=req.responseText, http_status=req.status, headers=req.getAllResponseHeaders())
-        res = jsonlib.loads(req.response)
-        if 'message' in res.keys():
-            raise CohereError(message=res['message'], http_status=req.status, headers=req.getAllResponseHeaders())
-        return res
-
     def __request(self, endpoint, json=None, method='POST') -> Any:
         headers = {
             'Authorization': 'BEARER {}'.format(self.api_key),
             'Content-Type': 'application/json',
             'Request-Source': self.request_source,
         }
-        if self.cohere_version != '':
-            headers['Cohere-Version'] = self.cohere_version
 
-        url = urljoin(self.api_url, endpoint)
-        if use_xhr_client:
-            response = self.__pyfetch(url, headers, jsonlib.dumps(json), method=method)
-            self.__print_warning_msg(response)
-            return response
-        else:
-            with requests.Session() as session:
-                retries = Retry(
-                    total=self.max_retries,
-                    backoff_factor=0.5,
-                    allowed_methods=['POST', 'GET'],
-                    status_forcelist=[429, 500, 502, 503, 504],
+        url = posixJoin(self.api_url, self.cohere_version, endpoint)
+        with requests.Session() as session:
+            retries = Retry(
+                total=self.max_retries,
+                backoff_factor=0.5,
+                allowed_methods=['POST', 'GET'],
+                status_forcelist=[429, 500, 502, 503, 504],
+            )
+            session.mount('https://', HTTPAdapter(max_retries=retries))
+            session.mount('http://', HTTPAdapter(max_retries=retries))
+
+            response = session.request(method, url, headers=headers, json=json, **self.request_dict)
+            try:
+                res = response.json()
```

Comment:
Can we do something like:

```python
try:
  response = session.request(method, url, headers=headers, json=json, **self.request_dict)
except:
  raise CohereConnectionError(...)

try:
  response.raise_for_status()
except HTTPError as e:
            code = e.response.status_code
            if code >= 500:
              raise CohereServerError(...)
            elif code >= 400
              raise CohereClientError(...)

res = response.json()
```

Reply:
Tried something like this, but the exceptions raised by the different libraries are quite byzantine